File: gblib/src/main/java/com/zwsi/gblib/GBStar.java
// Copyright 2018 Louis Perrochon. All rights reserved

// GBStar deals with anything on the star system wide level


package com.zwsi.gblib;

public class GBStar {

    private int nameIdx;
    private String name; // name of this system

    public int getIndex() { return index; }
    int index; // which position in Universe's star array

    private int numberOfPlanets = 2; // how many Planets in this solar Systems


    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    private int x; // x coordinate
    private int y; // y coordinate

    GBPlanet[] planetsArray = new GBPlanet[numberOfPlanets]; // the solar Systems // TODO make private an return copy in getter

    GBStar(GBData data) {
        nameIdx = GBData.Companion.selectStarNameIdx();
        name = GBData.Companion.starNameFromIdx(nameIdx);
        makePlanets();

        GBDebug.l3("Made System " + name);

        int[] coordinates = GBData.Companion.getStarCoordinates();
        x = coordinates[0];
        y = coordinates[1];

    }

    public String getName() {
        return name;
    }

    void consoleDraw(GBData data) {
        System.out.println("\n  " + "====================");
        System.out.println("  " + name + " System");
        System.out.println("  " + "====================");
        System.out.println("  The " + name + " system contains " + numberOfPlanets + " planet(s).");

        for (GBPlanet i : planetsArray) {
            if (i != null) {
                i.consoleDraw(data);
            }
        }

    }

    private void makePlanets() {
        GBDebug.l3("Making Planets for star " + name);

        for (int i = 0; i < planetsArray.length; i++) {
            planetsArray[i] = new GBPlanet();
            planetsArray[i].index = i;
        }

    }
}

File: gblib/src/main/java/com/zwsi/gblib/GBStar.kt
// Copyright 2018-2019 Louis Perrochon. All rights reserved

// GBStar deals with system wide level

package com.zwsi.gblib

import com.squareup.moshi.JsonClass
import com.zwsi.gblib.GBController.Companion.u
import kotlin.collections.HashSet

@JsonClass(generateAdapter = true)
data class GBStar(val uid: Int, val numberOfPlanets: Int, var loc: GBLocation) {

    var name: String

    init {
        name = GBData.starNameFromIdx(GBData.selectStarNameIdx())
        GBLog.d("Made System $name at location (${loc.x},${loc.y})")
    }

    // Planets
    // PERF ?? Don't keep materialized view, but compute on load and cache. Save happens every turn, but re-loads are rare.
    // Smaller file. Likely not worth it.
    var starUidPlanets: MutableSet<Int> = HashSet<Int>() // UID of planets. Persistent

    internal val starPlanets: List<GBPlanet>
        // PERF ?? Likely bad candidate. Cache the list and only recompute if the hashcode changes. Which requires Death Stars
        get() = starUidPlanets.map { u.planet(it) }

    // Ships
    var starUidShips: MutableSet<Int> = HashSet<Int>() // UID of shipsData. Persistent

    internal val starShips: List<GBShip>
        // PERF ?? Likely good candidate. Cache the list and only recompute if the hashcode changes.
        // This list changes frequently on update and is read frequently when drawing ( 30 times/second ) (no updates)
        get() = starUidShips.map { u.ship(it) }

    fun consoleDraw() {
        println("\n  " + "====================")
        println("  $name System")
        println("  " + "====================")
        println("  The $name system contains ${starUidPlanets.size} planet(s) and ${starUidShips.size} ship(s).")

        for (i in starPlanets) {
            i.consoleDraw()
        }
    }
}
